Core/MMAPs: Adjust walkable climb and fix a lot of mmap raycast issues

jackpoz opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Adjust walkable climb as recast using walkableClimb also to find the poly from a position, giving priority to polys that are below the position but closer than walkableClimb.
- Fix a lot of mmap raycast issues
- Note that some of the fixes require to re-extract mmaps, even if the previous ones are still compatible

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:** Closes #  (insert issue tracker number)


**Tests performed:** (Does it build, tested in-game, etc.)
- .go xyz 3328.058838 -3476.440918 287.074432 533 1.060287
- .go xyz 3175.989990 -3469.872070 287.049133 533 1.698029
- .go xyz 1438.516113 -4422.130371 25.462406 1 3.292355
- .go xyz -8461.000000 1012.000000 59.473511 0 5.808

**Known issues and TODO list:** (add/remove lines as needed)

- .go xyz 1027.279053 -4941.14502 12.63004 1 3.926178 (but works for me)
- .go xyz 1293.80725 -4403.5991 26.312632 1 0.153907 (but works for me)
- .go xyz 1268.545 -4435.84 26.637 1 3.740760 (but works for me)


<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

jackpoz committed:
Core/MMAPs: Adjust walkable climb

Adjust walkable climb as recast using walkableClimb also to find the poly from a position, giving priority to polys that are below the position but closer than walkableClimb.

Faq commented:
Any exact cases where this can be tested?

Ovahlord commented:
Naxxramas, Anub'Rekhan's spawn position. Try to blink into the left wall behind him. With the current checks you should end up not moving at all because in the current state, the path will go out of bounds, resulting in the vmap checks blocking the path to not fall out of the map

jackpoz committed:
Temporarily disable static collision (the whole check should be removed)

Ovahlord committed:
Core/Spells: removed deprecated mmap path check for TARGET_DEST_CASTER_FRONT_LEAP

Ovahlord committed:
Core/Objects: allow flying units to use the helper as well (flying units casting radius based spells)

jackpoz committed:
Code refactor

jackpoz committed:
Handle raycasts that end in a point with no height in the mmap mesh as PATHFIND_NOPATH

jackpoz commented:
To check: why the hitpoint sometimes has no height in the mmap mesh

jackpoz committed:
Walk back a bit from raycast hitpoints as sometime the 2D result point is outside of the polygons due to floating point errors.

jackpoz commented:
Found the reason: floating point precision issues

jackpoz committed:
Remove whitespace

jackpoz commented:
@Ovahlord do we even need anymore a full path when using raycast, even in case we hit no wall ? https://github.com/TrinityCore/TrinityCore/commit/4a197ba22af4eed01be632ea2dd7d103a963f043 tried to create a point path but failed retrieving the Z from mmaps, causing the issue you saw on Stormwind bridge with points under the bridge.

One could call getPolyHeight() for each point trying to guess which poly is the current one (or iterating them, even if they might not be in order) or just drop the whole "build a point path for raycast" thing.

I checked the sources and I didn't find any other usage of mmaps raycast other than the one we do.

jackpoz commented:
- [ ] To check: cases where raycast doesn't hit anything but height of endPos cannot be retrieved

jackpoz committed:
Revert 4a197ba22af4eed01be632ea2dd7d103a963f043 as a raycast point path should have the Z retrieved with getPolyHeight(). Raycast will only return a 2-point path with Start and Hitpoint/End

jackpoz committed:
Cleanup PathGenerator raycast case

jackpoz committed:
Fix PathGenerator raycast broken if start and end are on same poly.
Fix PathGenerator raycast broken if no wall is hit.
Remove unused case of using raycast with an existing previous path (can be added back properly if needed).
Remove forcing poly length to 2 when we actually have already the right number.

jackpoz committed:
Use closestPointOnPolyBoundary on the second try of finding a point on poly for raycast.
Note that in this case the mesh height is not used which might cause issues. The poly boundary height will be used instead.

jackpoz committed:
Handle cases where getPolyHeight() fails because the point is on polygon border (and caused by floating point imprecision)

jackpoz committed:
Add far from poly flags

jackpoz committed:
Set PATHFIND_INCOMPLETE in raycast case if startFarFromPoly or endFarFromPoly

jackpoz commented:
This PR is ready for testing

jackpoz commented:
PR hosted at https://tcubuntu.northeurope.cloudapp.azure.com/test/

jackpoz commented:
To fix:
- .go xyz 3175.989990 -3469.872070 287.049133 533 1.698029 case where endPoly is invalid (but raycast doesn't need endPoly

jackpoz committed:
Fix blink close to walls with no valid polygon behind the wall

jackpoz commented:
Fixed

jackpoz commented on src/server/game/Entities/Object/Object.cpp:
```diff
@@ -3285,7 +3286,9 @@ void WorldObject::MovePositionToFirstCollision(Position &pos, float dist, float
     float halfHeight = GetCollisionHeight() * 0.5f;
-    bool col = VMAP::VMapFactory::createOrGetVMapManager()->getObjectHitPos(GetMapId(),
+    bool col;
+    /*
```
this can stay commented for a while in case we have to restore it

Killyana commented:
Unable to blink in:  1027.279053, -4941.14502, 12.63004, 3.926178 even if you can walk on this path.

Jildor commented:
LK encounter platform
```
.go xyz 493.110474 -2120.135742 840.857117 631
```
use blink spellid: 1953
without this PR you don't move with blink
with this PR seems works fine
(same in Trial of the crusader and Malygos platforms)

Jildor commented:
.go xyz -4884.618164 -1956.716309 89.984360 1
get fear (.aura 5782)
without PR you go to the hell
wirh PR seems more accurate (you don't move alot, but you can't fall easy xD)

jackpoz commented:
@Killyana .go xyz 1027.279053 -4941.14502 12.63004 1 3.926178 worked for me on the test realm with a Human mage

Jildor commented:
Maybe add a new version of mmaps to ensure people re-extract mmaps can be better

jackpoz committed:
Require to re-extract mmaps

jackpoz commented:
@Ovahlord did you have anything else to change before we can merge this PR ? you can test it at https://tcubuntu.northeurope.cloudapp.azure.com/test/

Jildor commented:
seems with this PR fixes this issue: #24628

jackpoz commented:
nice
